### PR TITLE
Add information about rows distribution per task to exchange analyze

### PR DESCRIPTION
**Pull request by maciejgrzybek**
This will help identify data skew on repartitioning (remote exchanges).

Example output:
```
presto:default> explain analyze select sum(nationkey) over(partition by part) from foo;
                                                                        Query Plan
----------------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 1 [HASH]
     Cost: CPU 105.81ms, Input: 26 rows (598B). Per task: avg.: 13.0 std.dev.: 12.0, Output: 26 rows (234B)
     Output layout: [sum]
     Output partitioning: SINGLE []
     - Project[] => [sum:bigint]
             Cost: 4.52%, Input: 26 rows (832B), Output: 26 rows (234B), Filtered: 0.00%
             Input avg.: 0.81 lines, Input std.dev.: 535.10%
         - Window[partition by (part)][$hashvalue] => [nationkey:bigint, part:integer, $hashvalue:bigint, sum:bigint]
                 Cost: 44.72%, Output: 26 rows (832B)
                 Input avg.: 0.81 lines, Input std.dev.: 535.10%
                 sum := sum("nationkey")
             - LocalExchange[HASH][$hashvalue] ("part") => nationkey:bigint, part:integer, $hashvalue:bigint
                     Cost: 46.73%, Output: 26 rows (598B)
                     Input avg.: 0.81 lines, Input std.dev.: 535.10%
                 - RemoteSource[2] => [nationkey:bigint, part:integer, $hashvalue_5:bigint]
                         Cost: 4.02%, Output: 26 rows (598B)
                         Input avg.: 0.81 lines, Input std.dev.: 535.10%

 Fragment 2 [SOURCE]
     Cost: CPU 2.41s, Input: 26 rows (598B). Per task: avg.: 13.0 std.dev.: 12.0, Output: 26 rows (598B)
     Output layout: [nationkey, part, $hashvalue_6]
     Output partitioning: HASH [part][$hashvalue_6]
     - ScanProject[table = hive:hive:default:foo, originalConstraint = true] => [nationkey:bigint, part:integer, $hashvalue_6:bigint]
             Cost: 100.00%, Input: 26 rows (1.33kB), Output: 26 rows (598B), Filtered: 0.00%
             Input avg.: 13.00 lines, Input std.dev.: 92.31%
             $hashvalue_6 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("part"), 0))
             LAYOUT: hive
             nationkey := HiveColumnHandle{clientId=hive, name=nationkey, hiveType=bigint, hiveColumnIndex=0, columnType=REGULAR, comment=Optional.empty}
             part := HiveColumnHandle{clientId=hive, name=part, hiveType=int, hiveColumnIndex=4, columnType=REGULAR, comment=Optional.empty}


(1 row)
```

Please mind *Input: 26 rows (598B). Per task: avg.: 13.0 std.dev.: 12.0* in the beginning of `Fragment 1` description.

This is an output for a table with 26 rows where 25 had the same value of `part` column and only 1 had different value. There were 2 workers + 1 coordinator used to execute that query, so we ended up with 2 task collecting input from Fragment 2 (`hash-partition-count` was set to `100`).

CC: @KBP-TDC

### CODE REVIEW

**Comment by nezihyigitbasi** [comment]
Std dev is not easy to interpret, how about we add min and max (I guess we can also keep std dev)?

**Comment by maciejgrzybek** [comment]
Amended commit message:
s/Add information about rows distribution per task to *exchange* analyze/Add information about rows distribution per task to *explain* analyze/ ;)

**Comment by maciejgrzybek** [comment]
@nezi : std dev is really important because min/max solely won't tell you if you have a skew.
Imagine 100 nodes, 98 of them evenly distributed (around the same size of partition) but 1 received 0 rows and another gazillion rows.
min/max would tell you there's a big problem, while in fact there's none. I can add min/max to it if you see good use case for it but I oppose removing std dev.

**Comment by nezihyigitbasi** [comment]
@maciejgrzybek yeah I know, I don't think we should remove std dev as I said we can keep it in my initial comment. But, by only looking at std dev it's really hard to understand whether there's a problem or not (e.g., is std dev of 5 good or bad?) Ideally we should give a reasonable summary of the distribution (min, q1, mean, median, q3, max).

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java [hunk 1]
```diff
@@ -204,21 +204,29 @@ public static String textDistributedPlan(SubPlan plan, Metadata metadata, Sessio
...
+            double sdAmongTasks = Math.sqrt(1.0 / stageInfo.get().getTasks().size() * squaredDifferences);
+
             builder.append(indentString(1))
-                    .append(format("Cost: CPU %s, Input: %s (%s), Output: %s (%s)\n",
-                            stageStats.get().getTotalCpuTime(),
-                            formatPositions(stageStats.get().getProcessedInputPositions()),
-                            stageStats.get().getProcessedInputDataSize(),
-                            formatPositions(stageStats.get().getOutputPositions()),
-                            stageStats.get().getOutputDataSize()));
+                    .append(format("Cost: CPU %s, Input: %s (%s); per task: avg.: %s std.dev.: %s, Output: %s (%s)\n",
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java]: maybe drop `per task` information? I have an impression that either reader will not need this hint or he will not understand it.

**Comment by maciejgrzybek** [comment]
@nezihyigitbasi: would an output similar to the one proposed in https://github.com/prestodb/presto/pull/7661 be better instead of what is here?
E.g.  `{ count: 16, total: 60175, min: 0, p05: 0, p25: 0, p50: 0, p75: 8616, p95: 17192, max: 17192 }`
I know @sopel39 doesn't like it, claiming it's too verbose and not providing too much useful information.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java (hunk 1 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java]: User should be aware what `task` is. This concept is used in web UI already (in StagePerformance tab) . I'd leave it as is. We can change it in the future if needed (if users complain about it).

**Comment by maciejgrzybek** [comment]
I rebased this on master to resolve conflicts.
@nezihyigitbasi: I think we can add a distribution in `verbose` mode, to not pollute output in "standard" case.
So, once https://github.com/prestodb/presto/pull/7661 is merged, I extend this.

So far, the output would look like this:
```
presto> explain analyze select * from tpch.tiny.nation;
                                                                     Query Plan
-----------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 1 [SOURCE]
     Cost: CPU 2.12ms, Input: 25 rows (2.67kB); per task: avg.: 25.0 std.dev.: 0.0, Output: 25 rows (2.67kB)
     Output layout: [nationkey, name, regionkey, comment]
...
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java (hunk 1 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java]: I also had a conversation with @KBP-TDC about it and he claims leaving "per task" is more reasonable.

Comment (KBP-TDC) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java]: Yes, I believe people using EXPLAIN ANALYZE are advanced user who will also look at Presto UI and be familiar with Tasks and Stages. I also suggested we add similar metrics introduced by this PR to Presto UI for consistency.

**Comment by maciejgrzybek** [comment]
@martint what do you think about it?

**Comment by maciejgrzybek** [comment]
@martint Gentle reminder.

**Comment by maciejgrzybek** [comment]
@losipiuk or @sopel39: can you have a look please?

**Comment by maciejgrzybek** [comment]
Rebased on the most recent master.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java [hunk 2]
```diff
@@ -216,21 +216,29 @@ public static String textDistributedPlan(SubPlan plan, Metadata metadata, CostCa
...
                 fragment.getId(),
                 fragment.getPartitioning()));
 
-        if (stageStats.isPresent()) {
+        if (stageInfo.isPresent()) {
+            StageStats stageStats = stageInfo.get().getStageStats();
+
+            double avgPositionsPerTask = stageInfo.get().getTasks().stream().mapToLong(task -> task.getStats().getProcessedInputPositions()).average().orElse(Double.NaN);
+            double squaredDifferences = stageInfo.get().getTasks().stream().mapToDouble(task -> Math.pow(task.getStats().getProcessedInputPositions() - avgPositionsPerTask, 2)).sum();
+            double sdAmongTasks = Math.sqrt(1.0 / stageInfo.get().getTasks().size() * squaredDifferences);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java]: why not
```
Math.sqrt(squaredDifferences / stageInfo.get().getTasks().size())
```

**Comment by losipiuk** [comment]
See failing travis - probably unrelated but please ensure

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java (hunk 2 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java]: Doesn't matter. I think I may have not realized that `squaredDifferences` is of type `double`, thus don't have to enforce double arithmetics for division, or just without a reason? ;)

**Comment by maciejgrzybek** [comment]
@losipiuk: Travis is green.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java [hunk 3]
```diff
@@ -216,21 +216,29 @@ public static String textDistributedPlan(SubPlan plan, Metadata metadata, CostCa
...
+            double sdAmongTasks = Math.sqrt(squaredDifferences / stageInfo.get().getTasks().size());
+
             builder.append(indentString(1))
-                    .append(format("CPU: %s, Input: %s (%s), Output: %s (%s)\n",
-                            stageStats.get().getTotalCpuTime(),
-                            formatPositions(stageStats.get().getProcessedInputPositions()),
-                            stageStats.get().getProcessedInputDataSize(),
-                            formatPositions(stageStats.get().getOutputPositions()),
-                            stageStats.get().getOutputDataSize()));
+                    .append(format("CPU: %s, Input: %s (%s); per task: avg.: %s std.dev.: %s, Output: %s (%s)\n",
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java]: Maybe make the line more explicit, e.g:
```
... Input per task: avg: %s, std.dev: %s ...
```
It's not obvious that per task stats are input or output stats

Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java]: Where would you put info about total input?
Now it's
`Input: TOTAL; per task: avg: AVG, std.dev.: SD`

Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java]: OK, Let's leave it as it is. I was confused at first, but then I think it's fine.

**Pull request merged**